File: matriz_dinamica_python/matriz.py

class NodoCabecera:
    def __init__(self, indice):
        self.dato = indice
        self.sig= None
        self.ant=None
        self.listaInterna= listaDatos()

class NodoInterno:
    def __init__(self, dato,valorFila):
        self.dato = dato
        self.valorFila = valorFila
        self.sig = None
        self.ant = None

class matriz:
    def __init__(self):
        self.Columnas = ListaCabecera()

    def insertar(self,col,fila,valor):
        #buscar si la cabecera existe
        cabecera = self.Columnas.buscar(col)
         #si no existe creamos la cabecera
        if cabecera == None:
            cabecera = self.Columnas.insertar(col)

        #insertamos el dato a la lista de la cabecera
        cabecera.listaInterna.insertar_datos(valor,fila)

        return None

    def mostrar(self):
        self.Columnas.mostrar()


class ListaCabecera:
    def __init__(self):
        self.primero = None

    def insertar(self, dato):
        nuevo = NodoCabecera(dato)
        if self.primero == None:
            self.primero = nuevo
            return nuevo
        else:
            if nuevo.dato < self.primero.dato:
                nuevo.sig = self.primero
                self.primero.ant = nuevo
                self.primero = nuevo
                return nuevo
            else:
                if(self.primero.sig == None):
                    nuevo.ant = self.primero
                    self.primero.sig = nuevo
                    return nuevo
                else:
                    aux = self.primero
                    while(aux != None):
                        if nuevo.dato < aux.dato:
                            nuevo.sig = aux
                            nuevo.ant = aux.ant
                            aux.ant.sig = nuevo
                            aux.ant = nuevo
                            return nuevo
                        else:
                            aux = aux.sig
                
                return None

    def buscar(self, dato):
        aux = self.primero
        while aux != None:
            if(aux.dato == dato):
                return aux
            else: 
                aux = aux.sig
        
        return None

    def mostrar(self):
        aux = self.primero
        while aux != None:
            print(aux.dato)
            aux.listaInterna.mostrar_datos()
            aux = aux.sig
        
        return None

    
class listaDatos:
    def __init__(self):
        self.primero = None

    def insertar_datos(self, dato,fila):
        nuevo = NodoInterno(dato,fila)
        if self.primero == None:
            self.primero = nuevo
        else:
            if nuevo.valorFila < self.primero.valorFila:
                nuevo.sig = self.primero
                self.primero.ant = nuevo
                self.primero = nuevo
            else:
                if(self.primero.sig == None):
                    nuevo.ant = self.primero
                    self.primero.sig = nuevo
                    return nuevo
                else:
                    aux = self.primero
                    while(aux != None):
                        if nuevo.valorFila < aux.valorFila:
                            nuevo.sig = aux
                            nuevo.ant = aux.ant
                            aux.ant.sig = nuevo
                            aux.ant = nuevo
                            break
                        else:
                            aux = aux.sig
                
                return None
    
    def mostrar_datos(self):
        aux = self.primero
        while aux != None:
            print("     dato= "+str(aux.dato)+", fila: "+str(aux.valorFila))
            aux = aux.sig
        
        return None


m = matriz()

m.insertar(5,5,3)
m.insertar(1,5,7)
m.insertar(3,2,8)
m.insertar(1,1,9)
m.mostrar()




'''listaPrueba = ListaCabecera()
listaPrueba.insertar(5)
listaPrueba.insertar(10)
listaPrueba.insertar(2)
listaPrueba.mostrar()

pruebita = listaPrueba.buscar(2)
print(pruebita.dato) 

listadatos = listaDatos()
listadatos.insertar_datos(5,2)
listadatos.insertar_datos(8,3)
listadatos.insertar_datos(10,1)
listadatos.mostrar_datos() '''

File: matriz_dobleLista_javascritp/matriz.js
class nodo_interno{
    constructor(valor,x,y){
        this.valor = valor;
        this.x = x;
        this.y = y;
        //apuntadores
        this.sig = null;
        this.ant = null;

        this.arriba = null;
        this.abajo = null;
    }
}

class lista_interna{
    constructor(){
        this.primero = null;
    }

    insertar_x(valor, x,y){ //para las X usamos sig y ant, y el valor para compara y ordenar es Y
        let nuevo = new nodo_interno(valor,x,y);

        if(this.primero == null){
            this.primero = nuevo;
        }else{
            if(nuevo.y < this.primero.y){
                nuevo.sig = this.primero;
                this.primero.ant = nuevo;
                this.primero = nuevo;
            }else{
                let aux = this.primero;
                while(aux != null){
                    if(nuevo.y < aux.y){
                        nuevo.sig = aux;
                        nuevo.ant = aux.ant;
                        aux.ant.sig = nuevo;
                        aux.ant= nuevo;
                        break;
                    }else if(nuevo.x == aux.x && nuevo.y == aux.y){
                        console.log("La posicion ya esta ocupada-> "+nuevo.x+","+nuevo.y);
                        break;
                    }else{
                        if(aux.sig ==null){
                            aux.sig=nuevo;
                            nuevo.ant = aux;
                            break;
                        }else{
                            aux = aux.sig;
                        }
                    }
                }
            }
        }
    }

    insertar_y(valor, x,y){ //para las Y usamos arriba y abajo, y el valor para compara y ordenar es X
        let nuevo = new nodo_interno(valor,x,y);

        if(this.primero == null){
            this.primero = nuevo;
        }else{
            if(nuevo.x < this.primero.x){
                nuevo.abajo = this.primero;
                this.primero.arriba = nuevo;
                this.primero = nuevo;
            }else{
                let aux = this.primero;
                while(aux != null){
                    if(nuevo.x < aux.x){
                        nuevo.abajo = aux;
                        nuevo.arriba = aux.arriba;
                        aux.arriba.abajo = nuevo;
                        aux.arriba= nuevo;
                        break;
                    }else if(nuevo.x == aux.x && nuevo.y == aux.y){
                        console.log("La posicion ya esta ocupada-> "+nuevo.x+","+nuevo.y);
                        break;
                    }else{
                        if(aux.abajo ==null){
                            aux.abajo=nuevo;
                            nuevo.arriba = aux;
                            break;
                        }else{
                            aux = aux.abajo;
                        }
                    }
                }
            }
        }
    }

    recorrer_x(){
        let aux = this.primero;
        while(aux != null){
            console.log("valor =",aux.valor," - x = ",aux.x , " y = ",aux.y);
            aux = aux.sig;
        }
    }
    recorrer_y(){
        let aux = this.primero;
        while(aux != null){
            console.log("valor =",aux.valor," - x = ",aux.x , " y = ",aux.y);
            aux = aux.abajo;
        }
    }
}

//**************************** CABECERAS ************************/
class nodo_cabecera{
    constructor(dato){
        this.dato = dato;
        this.sig= null;
        this.ant = null;
        this.lista_interna = new lista_interna();
    }
}

class lista_cabecera{
    constructor(){
        this.primero = null;
    }

    insertar_cabecera(nuevo){

        if(this.primero == null){
            this.primero = nuevo;
        }else{
            if(nuevo.dato<this.primero.dato){
                nuevo.sig = this.primero;
                this.primero.ant=nuevo;
                this.primero = nuevo;
            }else{
                let aux = this.primero;
                while(aux != null){
                    if(nuevo.dato < aux.dato){
                        nuevo.sig = aux;
                        nuevo.ant = aux.ant;
                        aux.ant.sig = nuevo;
                        aux.ant = nuevo;
                        break;
                    }else{
                        if(aux.sig == null){
                            aux.sig = nuevo;
                            nuevo.ant = aux;
                            break;
                        }else{
                            aux = aux.sig;
                        }
                    }
                }
            }
        }
    }

    buscar_cabecera(dato){
        let aux = this.primero;
        while(aux != null){
            if(aux.dato == dato){
                return aux;
            }else{
                aux = aux.sig;
            }
        }
        return null;
    }

    recorrer(){
        let aux = this.primero;
        while(aux != null){
            console.log("dato =",aux.dato);
            aux = aux.sig;
        }
    }
}

//**************************** Matriz ************************/
class matriz{
    constructor(){
        this.cabecetas_x = new lista_cabecera();
        this.cabecetas_y = new lista_cabecera();
    }

    insertar(valor,x,y){
        let nodo_cabecera_X = this.cabecetas_x.buscar_cabecera(x);
        let nodo_cabecera_y = this.cabecetas_y.buscar_cabecera(y);

        if(nodo_cabecera_X == null){
            nodo_cabecera_X =  new nodo_cabecera(x);
            this.cabecetas_x.insertar_cabecera(nodo_cabecera_X);
        }

        if(nodo_cabecera_y == null){
            nodo_cabecera_y =  new nodo_cabecera(y);
            this.cabecetas_y.insertar_cabecera(nodo_cabecera_y);
        }

        //insertar en cabecera X
        nodo_cabecera_X.lista_interna.insertar_x(valor,x,y);
        //insertar en cabecera Y
        nodo_cabecera_y.lista_interna.insertar_y(valor,x,y);
    }

    recorrer_matriz(){
        console.log("cabeceras en X");
        let aux = this.cabecetas_x.primero;
        while(aux != null){
            console.log("   pos->"+aux.dato);
            let aux2 = aux.lista_interna.primero;
            while(aux2!= null){
                console.log("       -"+aux2.valor);
                aux2 = aux2.sig;
            }
            aux = aux.sig;
        }

        console.log("cabeceras en Y");
        aux = this.cabecetas_y.primero;
        while(aux != null){
            console.log("   pos->"+aux.dato);
            let aux2 = aux.lista_interna.primero;
            while(aux2!= null){
                console.log("       -"+aux2.valor);
                aux2 = aux2.abajo;
            }
            aux = aux.sig;
        }
    }

    graficar_matriz(){
        let cadena="";
        cadena+= "digraph Matriz{ \n";
        cadena+= "node[shape = box,width=0.7,height=0.7,fillcolor=\"azure2\" color=\"white\" style=\"filled\"];\n";
        cadena+= "edge[style = \"bold\"]; \n"
        //graficar el nodo matriz
        cadena+="node[label = Matriz fillcolor=\" darkolivegreen1\" pos = \"-1,1!\"]principal;"
        //graficar cabeceras X
        let aux_x = this.cabecetas_x.primero;
        while(aux_x!=null){
            cadena+="node[label = "+aux_x.dato+" fillcolor=\" azure1\" pos = \""+aux_x.dato+",1!\"]x"+aux_x.dato+";\n"
            aux_x = aux_x.sig;
        }
        aux_x = this.cabecetas_x.primero;
        while(aux_x.sig != null){
            cadena+="x"+aux_x.dato+"->"+"x"+aux_x.sig.dato+";\n"
            cadena+="x"+aux_x.sig.dato+"->"+"x"+aux_x.dato+";\n"
            aux_x = aux_x.sig;
        }

        if(this.cabecetas_x.primero!= null){
            cadena+="principal->x"+this.cabecetas_x.primero.dato+";\n";
        }
        //graficar cabeceras Y
        let aux_y = this.cabecetas_y.primero;
        while(aux_y!=null){
            cadena+="node[label = "+aux_y.dato+" fillcolor=\" azure1\" pos = \"-1,-"+aux_y.dato+"!\"]y"+aux_y.dato+";\n"
            aux_y = aux_y.sig;
        }
        aux_y = this.cabecetas_y.primero;
        while(aux_y.sig != null){
            cadena+="y"+aux_y.dato+"->"+"y"+aux_y.sig.dato+";\n"
            cadena+="y"+aux_y.sig.dato+"->"+"y"+aux_y.dato+";\n"
            aux_y = aux_y.sig;
        }

        if(this.cabecetas_x.primero!= null){
            cadena+="principal->y"+this.cabecetas_y.primero.dato+";\n";
        }
        //graficar nodos internos
        aux_x = this.cabecetas_x.primero;
        while(aux_x!=null){ //recorrer listas de x para graficar los nodos de sus lista interna
            let aux = aux_x.lista_interna.primero;
            while(aux!=null){
                cadena+="   node[label = "+aux.valor+" fillcolor=\" gold2\" pos = \""+aux.x+",-"+aux.y+"!\"]x"+aux.x+"y"+aux.y+";\n"
                aux = aux.sig;
            }

            //graficar flechitas
            aux = aux_x.lista_interna.primero;
            while(aux.sig!= null){
                cadena+="   x"+aux.x+"y"+aux.y+"->x"+aux.sig.x+"y"+aux.sig.y+";\n";
                cadena+="   x"+aux.sig.x+"y"+aux.sig.y+"->x"+aux.x+"y"+aux.y+";\n";
                aux= aux.sig;
            }
            if(aux_x.lista_interna.primero!= null){
                cadena+="x"+aux_x.dato+"->"+"x"+aux_x.lista_interna.primero.x+"y"+aux_x.lista_interna.primero.y+";\n";
            }

            aux_x = aux_x.sig;
        }

        aux_y = this.cabecetas_y.primero;
        while(aux_y!=null){ //recorrer la lista de y para graficar cada lista
            //graficar flechitas Y
            let aux = aux_y.lista_interna.primero;
            while(aux.abajo!= null){
                cadena+="   x"+aux.x+"y"+aux.y+"->x"+aux.abajo.x+"y"+aux.abajo.y+";\n";
                cadena+="   x"+aux.abajo.x+"y"+aux.abajo.y+"->x"+aux.x+"y"+aux.y+";\n";
                aux= aux.abajo;
            }
            if(aux_y.lista_interna.primero!= null){
                cadena+="y"+aux_y.dato+"->"+"x"+aux_y.lista_interna.primero.x+"y"+aux_y.lista_interna.primero.y+";\n";
            }
            aux_y = aux_y.sig;
        }

        cadena+= "\n}"
        console.log(cadena);
    }
}


let matriz1 = new matriz();

matriz1.insertar(0,0,0);
matriz1.insertar(50,0,1);
matriz1.insertar(5,1,1);
matriz1.insertar(6,2,3);
matriz1.insertar(1,10,1);
matriz1.insertar(2,1,2);
matriz1.insertar(7,3,3);

matriz1.recorrer_matriz();
matriz1.graficar_matriz();
